src/pip/_internal/cli/cmdoptions.py

```diff
@@ -920,6 +922,35 @@ def check_list_path_option(options):
 )  # type: Callable[..., Option]
 
 
+discovered_auth_plugins = {
+    entry_point.name: entry_point
+    for entry_point
+    in iter_entry_points('pip.auth_plugins')
+}
+
+
+def _handle_auth_plugin(option, opt_str, value, parser):
+    # type: (Option, str, str, OptionParser) -> None
+    plugin = discovered_auth_plugins[value].load()
+    setattr(parser.values, option.dest, plugin.Auth)
+
+
+auth_plugin = partial(
+    PipOption,
+    '--auth', '--auth-plugin',
+    dest='auth_class',
+    type='choice',
+    metavar='plugin',
+    default=None,
```

Comment:
I think we might want to have `MultiDomainBasicAuth` here.

Reply:
I considered this. But didn't want to couple cli and network code.